format/midi/metaevents.go

```diff
@@ -0,0 +1,393 @@
+}
+
+func decodeCuePoint(d *decode.D) {
+	d.FieldStrFn("cue", vlstring)
+}
+
+func decodeProgramName(d *decode.D) {
+	d.FieldStrFn("program", vlstring)
+}
+
+func decodeDeviceName(d *decode.D) {
+	d.FieldStrFn("device", vlstring)
+}
+
+func decodeMIDIChannelPrefix(d *decode.D) {
+	d.FieldUintFn("channel", func(d *decode.D) uint64 {
+		channel := uint64(0)
+
+		if data, err := vlf(d); err != nil {
+			d.Errorf("%v", err)
+		} else {
+			for _, b := range data {
+				channel <<= 8
+				channel |= uint64(b & 0x00ff)
+			}
+		}
+
+		return channel
+	})
+}
+
+func decodeMIDIPort(d *decode.D) {
+	d.FieldUintFn("port", func(d *decode.D) uint64 {
+		port := uint64(0)
+
+		if data, err := vlf(d); err != nil {
+			d.Errorf("%v", err)
+		} else {
+			for _, b := range data {
+				port <<= 8
+				port |= uint64(b & 0x00ff)
+			}
+		}
+
+		return port
+	})
+}
+
+func decodeTempo(d *decode.D) {
+	d.FieldUintFn("tempo", func(d *decode.D) uint64 {
+		tempo := uint64(0)
+
+		if data, err := vlf(d); err != nil {
+			d.Errorf("%v", err)
+		} else {
+			for _, b := range data {
+				tempo <<= 8
+				tempo |= uint64(b & 0x00ff)
+			}
+		}
+
+		return tempo
+	})
+}
+
+func decodeSMPTEOffset(d *decode.D) {
+	d.FieldStruct("offset", func(d *decode.D) {
+		var data []uint8
+		var err error
+
+		d.FieldStrFn("bytes", func(d *decode.D) string {
+			if data, err = vlf(d); err != nil {
+				d.Errorf("%v", err)
+			} else {
+				return fmt.Sprintf("%v", data)
+			}
+
+			return "[]"
+		})
+
+		if len(data) > 0 {
+			d.FieldUintFn("framerate", func(d *decode.D) uint64 {
+				return uint64((data[0] >> 6) & 0x03)
+			}, framerates)
+
+			d.FieldValueUint("hour", uint64(data[0]&0x01f))
+		}
+
+		if len(data) > 1 {
+			d.FieldValueUint("minute", uint64(data[1]))
+		}
+
+		if len(data) > 2 {
+			d.FieldValueUint("second", uint64(data[2]))
+		}
+
+		if len(data) > 3 {
+			d.FieldValueUint("frames", uint64(data[3]))
+		}
+
+		if len(data) > 4 {
+			d.FieldValueUint("fractions", uint64(data[4]))
+		}
+	})
+}
+
+func decodeTimeSignature(d *decode.D) {
+	d.FieldStruct("signature", func(d *decode.D) {
+		var data []uint8
+		var err error
+
+		d.FieldStrFn("bytes", func(d *decode.D) string {
+			if data, err = vlf(d); err != nil {
+				d.Errorf("%v", err)
+			} else {
+				return fmt.Sprintf("%v", data)
+			}
+
+			return "[]"
+		})
+
+		if len(data) > 0 {
+			d.FieldValueUint("numerator", uint64(data[0]))
+		}
+
+		if len(data) > 1 {
+			denominator := uint64(1)
+			for i := uint8(0); i < data[1]; i++ {
+				denominator <<= 1
+			}
+
+			d.FieldValueUint("denominator", denominator)
+		}
+
+		if len(data) > 2 {
+			d.FieldValueUint("ticks_per_click", uint64(data[2]))
+		}
+
+		if len(data) > 3 {
+			d.FieldValueUint("thirty_seconds_per_quarter", uint64(data[3]))
+		}
+	})
+}
+
+func decodeKeySignature(d *decode.D) {
+	d.FieldUintFn("key", func(d *decode.D) uint64 {
+		key := uint64(0)
+
+		if data, err := vlf(d); err != nil {
+			d.Errorf("%v", err)
+		} else {
```

Comment:
Same encoding as in decodeSequenceNumber? if so make sene to refactor to share somehow?

Reply:
Fixed in[578b7e78e395f4c1baefd8adb2cd6794176f2d2b](https://github.com/wader/fq/pull/1004/commits/578b7e78e395f4c1baefd8adb2cd6794176f2d2b)